test/src/test/java/jenkins/model/Jenkins56575Test.java

```diff
@@ -0,0 +1,227 @@
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import com.gargoylesoftware.htmlunit.HttpMethod;
+import com.gargoylesoftware.htmlunit.Page;
+import com.gargoylesoftware.htmlunit.WebRequest;
+import com.gargoylesoftware.htmlunit.WebResponse;
+import hudson.Launcher;
+import hudson.Main;
+import hudson.lifecycle.Lifecycle;
+import hudson.util.StreamTaskListener;
+import org.apache.commons.io.FileUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.JenkinsRule.WebClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertEquals;
+
+public class Jenkins56575Test {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    private File jar;
+
+    private AtomicReference<Boolean> restartCalled = new AtomicReference<>(false);
+    private AtomicReference<Boolean> verifyRestartableCalled = new AtomicReference<>(false);
+    private Semaphore restartSemaphore = new Semaphore(0);
+
+    private Lifecycle previousLifecycle;
+    private Field instanceField;
+
+    @Before
+    public void setupMockLifecycle() throws Exception {
+        Main.isUnitTest = false;
+        j.jenkins.setCrumbIssuer(null);
+
+        previousLifecycle = Lifecycle.get();
+
+        instanceField = Lifecycle.class.getDeclaredField("INSTANCE");
+        instanceField.setAccessible(true);
+        instanceField.set(null, new Lifecycle() {
+            @Override
+            public void verifyRestartable() {
+                verifyRestartableCalled.set(true);
+            }
+
+            @Override
+            public void restart() {
+                restartCalled.set(true);
+                restartSemaphore.release();
+            }
+        });
+    }
+
+    @Before
+    public void grabCliJar() throws IOException {
+        jar = tmp.newFile("jenkins-cli.jar");
+        FileUtils.copyURLToFile(j.jenkins.getJnlpJars("jenkins-cli.jar").getURL(), jar);
+    }
+
+    @After
+    public void resetSettings() throws Exception {
+        Main.isUnitTest = true;
+
+        instanceField.set(null, previousLifecycle);
+    }
+
+    @Test
+    public void testRestart_regularClient() throws Exception {
+        WebClient wc = j.createWebClient()
+                .withRedirectEnabled(false)
+                .withThrowExceptionOnFailingStatusCode(false);
+        WebRequest request = new WebRequest(new URL(j.getURL() + "restart"), HttpMethod.POST);
```

Comment:
These four lines are roughly the same as the first four lines of the next block, perhaps refactor as a function that takes the `restart`/`restartSimple` action as an argument?

(I'm currently running through IntelliJ's tooling, and it complains about common code -- also some reviewers ding me for things like this....)

Reply:
Personally I prefer to duplicate lines in the test part if that keeps the readability higher.
I can simplify the call to `withThrowExceptionOnFailingStatusCode` as it's not strictly required and the request creation could be refactored. More simplification/refactor seems overengineering at this point. Is it fine for you?